### PR TITLE
Refactor reporting config file reading

### PR DESCRIPTION
**Pull request by jorblancoa**
CI_BRANCHES:NEURON_BRANCH=master

### CODE REVIEW

**Comment by jorblancoa** [comment]
retest this please jenkins

File: coreneuron/io/reports/report_configuration_parser.cpp [hunk 1]
```diff
@@ -59,62 +61,49 @@ enum class TargetType
...
-        nrn_abort(1);
-    }
+    std::ifstream report_conf(conf_file);
+    std::string line;
+    std::getline(report_conf, line);
 
-    fgets(raw_line, REPORT_MAX_FILEPATH_LEN, fp);
-    sscanf(raw_line, "%d\n", &num_reports);
+    std::istringstream iss(line);
+    int num_reports = 0;
```
Comment (sergiorg-hpc) [inline coreneuron/io/reports/report_configuration_parser.cpp:83]: `TargetType` is an `int` internally, why the change? Just curious, did it fail to compile or something?

File: coreneuron/io/reports/report_configuration_parser.cpp [hunk 2]
```diff
@@ -59,62 +61,49 @@ enum class TargetType
...
+                                                              std::string &spikes_population_name) {
     std::vector<ReportConfiguration> reports;
-    int num_reports = 0;
-    char report_on[REPORT_MAX_NAME_LEN] = "";
-    char raw_line[REPORT_MAX_FILEPATH_LEN] = "";
-    char spikes_population[REPORT_MAX_NAME_LEN] = "";
-    TargetType target_type;
-    int* gids;
+    std::string report_on;
+    int target_type;
```
Comment (sergiorg-hpc) [inline coreneuron/io/reports/report_configuration_parser.cpp:76]: `TargetType` is an `int` internally, why the change? Just curious, did it fail to compile or something?

(Clicked the wrong place, excuse me for re-adding the comment, hehe)

File: coreneuron/io/reports/report_configuration_parser.cpp [hunk 3]
```diff
@@ -59,62 +61,49 @@ enum class TargetType
...
-    token = strtok(nullptr, "\n");
-    if (!token) {
-        std::cerr << "Error : Invalid report variable, should be mch_name.var_name" << std::endl;
-        nrn_abort(1);
-    }
-    strcpy(config.var_name, token);
+void parse_filter_string(const std::string &filter, ReportConfiguration &config) {
+    std::istringstream iss(filter);
+    std::string token;
+    std::getline(iss, config.mech_name, '.');
```
Comment (sergiorg-hpc) [inline coreneuron/io/reports/report_configuration_parser.cpp:67]: Just curious, will this fail if the `.` is not found like the original code?

File: coreneuron/io/reports/report_configuration_parser.cpp [hunk 4]
```diff
@@ -59,62 +61,49 @@ enum class TargetType
...
 
-    FILE* fp = fopen(conf_file, "r");
-    if (!fp) {
-        std::cerr << "Cannot open configuration file: " << conf_file << ", aborting execution"
-                  << std::endl;
-        nrn_abort(1);
-    }
+    std::ifstream report_conf(conf_file);
+    std::string line;
+    std::getline(report_conf, line);
```
Comment (sergiorg-hpc) [inline coreneuron/io/reports/report_configuration_parser.cpp:80]: Why keeping the `std::getline` and not trying to read directly from the `std::ifstream`?

File: coreneuron/io/reports/report_handler.cpp [hunk 5]
```diff
@@ -11,7 +11,7 @@ void ReportHandler::create_report(double dt, double tstop, double delay) {
     }
     m_report_config.stop = std::min(m_report_config.stop, tstop);
 
-    m_report_config.mech_id = nrn_get_mechtype(m_report_config.mech_name);
+    m_report_config.mech_id = nrn_get_mechtype(m_report_config.mech_name.data());
```
Comment (sergiorg-hpc) [inline coreneuron/io/reports/report_handler.cpp:14]: Is it `.data()` or `.c_str()`? The code might expect a null-terminated string, so this could be dangerous, right?

File: coreneuron/io/reports/report_handler.cpp [hunk 6]
```diff
@@ -51,7 +51,7 @@ void ReportHandler::create_report(double dt, double tstop, double delay) {
         }
         if (!vars_to_report.empty()) {
             auto report_event =
-                std::make_unique<ReportEvent>(dt, t, vars_to_report, m_report_config.output_path);
+                std::make_unique<ReportEvent>(dt, t, vars_to_report, m_report_config.output_path.data());
```
Comment (sergiorg-hpc) [inline coreneuron/io/reports/report_handler.cpp:54]: See line 14, and replace in the other cases if needed.

**Comment by sergiorg-hpc** [comment]
Nice job, @jorblancoa, just minor comments or clarifications.

File: coreneuron/io/reports/report_handler.cpp (hunk 5 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_handler.cpp:14]: Actually it just converts the char* to string, so .data() would be just fine here!
```
int nrn_get_mechtype(const char* name) {
    std::string str(name);
```
And i cant just pass the string because the function is used in lots other places...

File: coreneuron/io/reports/report_handler.cpp (hunk 6 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_handler.cpp:54]: same!

File: coreneuron/io/reports/report_configuration_parser.cpp (hunk 3 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_configuration_parser.cpp:67]: I thought about that, but if it is not x.y format, neurodamus fails already

File: coreneuron/io/reports/report_configuration_parser.cpp (hunk 2 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_configuration_parser.cpp:76]: It didnt compile since istringstream >> didnt like the enum.. so i read the int and convert it after

File: coreneuron/io/reports/report_handler.cpp (hunk 5 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_handler.cpp:14]: Actually, from C++11 are equivalent

File: coreneuron/apps/main1.cpp [hunk 7]
```diff
@@ -407,9 +407,9 @@ static void trajectory_return() {
 
 std::unique_ptr<ReportHandler> create_report_handler(ReportConfiguration& config) {
     std::unique_ptr<ReportHandler> report_handler;
-    if (std::strcmp(config.format, "Bin") == 0) {
+    if (config.format == "Bin") {
```
[the remark this answers is not in the document]
Comment (jorblancoa) [reply coreneuron/apps/main1.cpp:410]: It could, but then there are a lot of places where we log the format string, then we would need convertors. So I am not sure if it is worth...

[the remark this answers is not in the document]
Comment (alkino) [reply coreneuron/apps/main1.cpp:410]: I don't have the answer. The worth is something more understandable than `std::string`, it depends how much user use it.

File: coreneuron/io/reports/report_configuration_parser.cpp (hunk 4 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_configuration_parser.cpp]: done!

File: coreneuron/io/reports/report_configuration_parser.cpp [hunk 8]
```diff
@@ -59,62 +61,44 @@ enum class TargetType
...
+                       report.type_str.begin(),
+                       [](unsigned char c) { return std::tolower(c); });
+        report.output_path = output_dir + "/" + report.name;
+        if (report.type_str == "compartment") {
+            if (report_on == "i_membrane") {
                 nrn_use_fast_imem = true;
                 report.type = IMembraneReport;
             } else {
-                switch (target_type) {
+                switch (TargetType(target_type)) {
```
Comment (alexsavulescu) [inline coreneuron/io/reports/report_configuration_parser.cpp:101]: use c++ casting

**Review by alexsavulescu — COMMENTED** [review]
LGTM, one minor nitpick

File: coreneuron/io/reports/report_configuration_parser.cpp (hunk 8 above)
Comment (jorblancoa) [reply coreneuron/io/reports/report_configuration_parser.cpp:101]: static_cast<TargetType>(target_type)?

Comment (alexsavulescu) [reply coreneuron/io/reports/report_configuration_parser.cpp:101]: Yes
```suggestion
                switch (static_cast<TargetType>(target_type)) {
```

**Review by sergiorg-hpc — APPROVED** [review]

**Comment by sergiorg-hpc** [comment]
Extended Simulation Stack, that includes the Section target tests, successfully passed:
https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/2781/pipeline/

**Review by alexsavulescu — APPROVED** [review]

**Review by pramodk — APPROVED** [review]
